### ISSUE TITLE
Start and stop sendmail queues for MailScanner 5.x

### ISSUE THREAD

**Issue by wolf340** [MailScanner/v5#23]
[mailscanner-sendmail.zip](https://github.com/MailScanner/v5/files/529452/mailscanner-sendmail.zip)

Put in /etc/init.d/ and activate using

chmod 755 /etc/init.d/mailscanner-sendmail
chkconfig -add mailscanner-sendmail 
chkconfig mailscanner-sendmail on

Wolfgang

**Comment by msapiro** [comment]
Script and instructions work for RedHat/CentOS, but not Debian/Ubuntu. For that and perhaps others, the script needs LSB info, something like the following and maybe Should-Start: and Should-Stop: as well.
```
### BEGIN INIT INFO
# Provides:          sendmail mail-transport-agent
# Required-Start:    $local_fs $remote_fs $syslog $named $network $time
# Required-Stop:     $local_fs $remote_fs $syslog $named $network
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: Sendmail Mail Transport Agent
# Description:       sendmail is a Mail Transport agent
### END INIT INFO
```

and activation is
```
update-rc.d mailscanner-sendmail defaults
```
and with systemd
```
systemctl enable mailscanner-sendmail
```

**Comment by odeneriksson** [comment]
Please add this:

--- mailscanner-sendmail	2016-10-12 22:35:56.000000000 +0200
+++ mailscanner-sendmail.oden	2016-11-02 12:26:33.654909142 +0100
@@ -65,8 +65,16 @@
                       -ODeliveryMode=queueonly \
                       -OQueueDirectory=$INQDIR \
                       -OPidFile=$INPID 
-        touch /var/run/sm-client.pid
-        chown $MSPUSER:$MSPGROUP /var/run/sm-client.pid 2>/dev/null
+
+	# rhbz#200920
+	if [ ! -f $SMPID ]; then
+	    touch $SMPID
+	    chown $MSPUSER:$MSPGROUP $SMPID 2>/dev/null
+	    if [ -x /usr/sbin/selinuxenabled ] && /usr/sbin/selinuxenabled; then
+		/sbin/restorecon $SMPID
+	    fi
+	fi
+
         $SENDMAIL -L sm-msp-queue -Ac -q15m -OPidFile=$SMPID 2>/dev/null
         success
         echo

**Comment by odeneriksson** [comment]
Please add this:

[mailscanner-sendmail.diff.txt](https://github.com/MailScanner/v5/files/566374/mailscanner-sendmail.diff.txt)

**Comment by jcbenton** [comment]
please post the full script

**Comment by jcbenton** [comment]
can we get a complete script that will work for both so I can add this? we will also need to update the installers and spec files.

**Comment by wolf340** [comment]
Complete script attached with proposed changes merged in. Note that I did not test the LSB info. Someone with an install on such an system should test that.
[mailscanner-sendmail.zip](https://github.com/MailScanner/v5/files/578584/mailscanner-sendmail.zip)

**Comment by wolf340** [comment]
> can we get a complete script that will work for both so I can add this? we
> will also need to update the installers and spec files.
>

complete script including proposed changes posted.

Wolfgang

**Comment by msapiro** [comment]
I can't actually test the script because I don't have a platform with MailScanner and sendmail to test on, but I did test installing and removing the script with `update-rc.d`, and for that purpose, the LSB INIT INFO looks fine.

**Comment by mgtremaine** [comment]
The attached script works fine on a Centos 5.11 install.. I just made the leap from mailscanner 4.85.2-3 to 5.0.3-7  the only issue I ran into was the installer barfed at the end and I grabbed the install.sh from git and re-ran it. Then I had to install this script to control sendmail. (Also the /var/spool/mqueue.in dir was blown away and i had to remake by hand)...

I'll be rolling out more install in the weeks to come (Centos 6 and 7). Thanks for the handwork!

-Mike

**Comment by ajos1** [comment]
Hmmm.. not sure if the latest script is FC25 compliant.. 

If I run the latest version of the script in various ways, this is what happens. The main issue is that "/sbin/service mailscanner-sendmail start" fails.. it just sits there for ages before coming up with "Job for mailscanner-sendmail.service failed because a timeout was exceeded."

 - - -


[root@rmsvajos203 init.d]# /sbin/service mailscanner-sendmail stop
Stopping mailscanner-sendmail (via systemctl):             [  OK  ]



[root@rmsvajos203 init.d]# /sbin/service mailscanner-sendmail start
Starting mailscanner-sendmail (via systemctl):  Job for mailscanner-sendmail.service failed because a timeout was exceeded.


[root@rmsvajos203 init.d]# sh mailscanner-sendmail stop
Stopping sendmail daemons:
         incoming sendmail:                                [  OK  ]
         outgoing sendmail:                                [  OK  ]


[root@rmsvajos203 init.d]# sh mailscanner-sendmail start
Starting sendmail daemons:
         incoming sendmail:                                [  OK  ]
         outgoing sendmail:                                [  OK  ]

**Comment by mikaku** [comment]
It looks like `systemd` don't like multiple *.pid* files, so just changing the line:

`# pidfile: /var/run/sendmail.*.pid, /var/run/sm-client.pid`

by this one

`# pidfile: /var/run/sm-client.pid`

does the trick.

**Comment by jcbenton** [comment]
Do we have a final version of this? I will include it if you can give me a finished working script.

**Comment by wolf340** [comment]
The only change to the version I posted on Nov 8 is the # pidfile line change suggested above. It still works fine with this change on RHEL6/CentOS6. Script has not been tested on SUSE, I think.

**Comment by jcbenton** [comment]
Please post the full file so we can have it for records.

**Comment by wolf340** [comment]
[mailscanner-sendmail.zip](https://github.com/MailScanner/v5/files/900514/mailscanner-sendmail.zip)

**Comment by jcbenton** [comment]
Ok, I am reviewing this script. RHEL specific? Lines like this:

`# Source function library.
. /etc/rc.d/init.d/functions`

need to look like this:

`# Source function library.
if [ -f '/etc/rc.d/init.d/functions' ]; then
. /etc/rc.d/init.d/functions
fi
`

It also needs the FreeBSD and Debian requirements in the comments. Welcome to the world of programming for everyone.

So, RHEL specific? I can include it but have to denote this. We would at some point have to update it to work across all the supported distributions.

**Comment by wolf340** [comment]
I think it has been successfully tested on Ubuntu/Debian as well (that's why we added those required comments), so not exactly RHEL specific

**Comment by jcbenton** [comment]
I doubt it. This does not exist on Debian.

`. /etc/sysconfig/network`

nor this:
`/sbin/portrelease `

nor this:
`/usr/sbin/selinuxenabled `

this is not used anymore in mailscanner:
`/var/lock/subsys`


I will leave this open. I would have to rewrite the whole thing, but the script you provide gives me a good start. Thank you.

**Comment by philrandal** [comment]
Might it not be better to have two separate 'services', mailscanner-sendmail-in and mailscanner-sendmail-out rather than to combine both as the same service?

**Comment by sadsfae** [comment]
For what it's worth I get the same timeouts as mentioned above with the latest attachment.

```
[root@funcamp ~]# systemctl status mailscanner-sendmail
● mailscanner-sendmail.service - LSB: Sendmail Mail Transport Agent
   Loaded: loaded (/etc/rc.d/init.d/mailscanner-sendmail; bad; vendor preset: disabled)
   Active: inactive (dead)
     Docs: man:systemd-sysv-generator(8)

Apr 26 16:48:08 funcamp.net sendmail[30222]: starting daemon (8.14.7): SMTP
Apr 26 16:48:08 funcamp.net sm-msp-queue[30228]: starting daemon (8.14.7): queueing@00:15:00
Apr 26 16:48:08 funcamp.net mailscanner-sendmail[30216]: incoming sendmail: [  OK  ]
Apr 26 16:48:08 funcamp.net sendmail[30232]: starting daemon (8.14.7): queueing@00:15:00
Apr 26 16:48:08 funcamp.net mailscanner-sendmail[30216]: outgoing sendmail: [  OK  ]
Apr 26 16:48:08 funcamp.net systemd[1]: PID file /var/run/sendmail.*.pid, /var/run/sm-client.pid not readable (yet?) after start.
```

I have a working CentOS7 MailScanner 5.0.3 and Sendmail 8.14.7 setup and I'm happy to test if it helps.

**Comment by mikaku** [comment]
@sadsfae, have you read my post right after his one?

**Comment by sadsfae** [comment]
> @sadsfae, have you read my post right after his one?

Yes, still no dice for me.  The systemd service still doesn't start.

**Comment by MortenCB** [comment]
I was really puzzled to why sendmail didn't start up on my new centos 7 server, as on my old centos 5 server.  Found this bug, and got it to work with the latest mailscanner-sendmail.zip linked above, and after changing the pid-line as someone pointed out.  This should really be fixed and included in the install.sh since MailScanner doesn't work as it is deployed now on centos/rhel 7.

**Comment by jcbenton** [comment]
Ok, I have recently seen the light on the simplicity of systemd. What I would recommend:

- remove the symlink /etc/init.d/mailscanner
- create a simple systemd style startup for mailscanner
- create a similar in/out for sendmail in systemd

I would do this for you guys, but I am swamped with Mailborder v5. (I like food and shelter.)

**Comment by shawniverson** [comment]
I will see what I can do for you.  Have some downtime today...

**Comment by shawniverson** [comment]
Started work on this.   Working on extending mailscanner.spec to detect CentOS and RHEL 7 and apply systemd specific configuration instead of init.  Working on systemd script for mailscanner.  I'll keep going with this.

**Comment by shawniverson** [comment]
Work in progress here  https://github.com/MailScanner/v5/pull/54

**Comment by shawniverson** [comment]
Resolved in v5.0.6-1
